docs: how to use pebble api (#545)
canonical/pebble: docs/how-to/use-the-pebble-api.md

# How to use the Pebble API to manage services

Pebble provides a command-line interface for managing services, similar to systemd. However, Pebble distinguishes itself with its dedicated REST API.

This guide demonstrates how to use the Pebble API to programmatically manage services as part of an automated workflow that starts, tests, and stops interdependent services in a CI pipeline. While shell scripts could be used for this purpose, the Pebble API offers a more robust approach with easier error handling and reusable components. This reduces manual work, errors, and ensures consistency across environments.

## Use the API

```{include} /reuse/api.md
   :start-after: Start: Pebble API overview
   :end-before: End: Pebble API overview
```

> For reference information about the API, see [](../explanation/api-and-clients) and [](../reference/api).

Reviewer: It's hard to tell in this context what the difference between these two links are. How about:

```suggestion
For an explanation of API access levels, see [](../explanation/api-and-clients). For the full API reference, see [](../reference/api).
```

Reviewer 2: Side question - Are we keeping [API and clients](https://canonical-pebble.readthedocs-hosted.com/en/latest/explanation/api-and-clients/) as-is? I thought we had intended to move some (all?) of it into [API](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/api/)

Reviewer: Oooh, yes, I think that's part of the confusion. I think that's a good idea. I think most of the introduction in "API and clients" is repeated in the API reference anyway. And the identities stuff is over in References > Identities. So the only thing that actually needs to move IMO is the "API access levels" section.

Probably best to do that in a separate PR, though.

Suppose we start the Pebble daemon with no default services and an empty layer:

```{terminal}
   :input: pebble run
2024-12-30T01:07:10.275Z [pebble] Started daemon.
2024-12-30T01:07:10.281Z [pebble] POST /v1/services 75.042µs 400
2024-12-30T01:07:10.281Z [pebble] Cannot start default services: no default services
```

We can then use a Python client to interact with Pebble:

```python
import ops

client = ops.pebble.Client("/path/to/.pebble.socket")

# get services
services = client.get_services()
assert len(services) == 0

# add layer
layerYAML = """
services:
  svc1:
    override: replace
    command: sleep 100
"""

> client.add_layer(label="ci", layer=layerYAML, combine=True)

Reviewer: `label` and `layer` are required positional args, so you'd typically use them without keywords. How about:
```suggestion
client.add_layer("label1", layerYAML, combine=True)
```

# start services
client.start_services(["svc1"])  # Python client also waits for change to finish

#  get services, the service svc1 should be active
services = client.get_services()
for svc in services:
    print(f"The status of service {svc.name} is: {svc.current}.")

# Now we can run some tests against those services.

# stop services
client.stop_services(["svc1"])  # Python client also waits for change to finish
```

You can also use Go or curl to achieve the same result. For more information, see {ref}`api_go_client` and {ref}`api_curl`.

## See more

- [Go client](https://pkg.go.dev/github.com/canonical/pebble/client)
- [Python client for Pebble API](https://ops.readthedocs.io/en/latest/reference/pebble.html)
- [API and clients](../explanation/api-and-clients)

> - [API](../reference/api)

Reviewer: Again, it's hard to tell what distinguishes these last two links. Can we do something like:

- [Explanation of API access control](...)
- [API reference](...)

# How to use the Pebble API to manage services

> Pebble provides a command-line interface (CLI) for managing services, similar to systemd. However, Pebble distinguishes itself with its dedicated REST API.

Reviewer 2: Being extremely picky, we could remove "CLI" as that term isn't used anywhere else in the page. I'm actually good either way.

```suggestion
Pebble provides a command-line interface for managing services, similar to systemd. However, Pebble distinguishes itself with its dedicated REST API.
```

> This guide demonstrates how to use the Pebble API to programmatically manage services as part of an automated workflow where we can start, test, and stop interdependent services in a CI pipeline. While shell scripts could be used for this purpose, Pebble API offers a more robust approach with easier error handling and reusable components. This reduces manual work, errors, and ensures consistency across environments.

Reviewer 2: Some small suggestions for reading flow:

```suggestion
This guide demonstrates how to use the Pebble API to programmatically manage services as part of an automated workflow that starts, tests, and stops interdependent services in a CI pipeline. While shell scripts could be used for this purpose, the Pebble API offers a more robust approach with easier error handling and reusable components. This reduces manual work, errors, and ensures consistency across environments.
```

> ## Using the API

Reviewer 2: I think this should be "Use the API", for consistency with section headings in other how-to guides:

```suggestion
## Use the API
```

```{include} /reuse/api.md
   :start-after: Start: Pebble API overview
   :end-before: End: Pebble API overview
```

For reference information about the API, see [](../explanation/api-and-clients) and [](../reference/api).

Suppose we start the Pebble daemon with no default services and an empty layer:

```{terminal}
   :input: pebble run
2024-12-30T01:07:10.275Z [pebble] Started daemon.
2024-12-30T01:07:10.281Z [pebble] POST /v1/services 75.042µs 400
2024-12-30T01:07:10.281Z [pebble] Cannot start default services: no default services
```

> Here is an example in Python:

Reviewer 2: I suggest making this a bit more descriptive of what follows. Maybe something like:

```suggestion
We can then interact with Pebble using a Python client:
```

> ```python

Reviewer 2: I have a question about the Python (maybe I didn't understand it correctly): Is `client.wait_change(changeID)` needed at the end? If not, should we remove the `changeID` variable too (on lines 46 and 56), or would that variable be useful for other things?

Author: This is an oversight, I will remove the wait change, add a comment instead, and also remove the variable.

## Use the API

> Pebble's API allows remote client interaction with the daemon. It uses HTTP over a Unix socket, with access controlled by user ID.

Reviewer: Prefer a more active voice:

```suggestion
Pebble's API allows clients to interact remotely with the daemon. It uses HTTP over a Unix socket, with access controlled by user ID.
```
